File: dp/leetcode_1239.java
package dp;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class leetcode_1239 {
    public static void main(String[] args) {
        leetcode_1239 lc = new leetcode_1239();

        List<String> arr = new ArrayList<>();
        arr.add("cha");
        arr.add("r");
        arr.add("act");
        arr.add("ers");

        System.out.println(lc.maxLength(arr));
    }

    public int maxLength(List<String> arr) {
        return helper(arr, 0, "");
    }

    private int helper(List<String> arr, int index, String cur) {
        if (index == arr.size()) return cur.length();

        int res = cur.length();
        for (int i = index; i < arr.size(); i++) {
            String str = cur + arr.get(i);
            if (isUnique(str)) {
                res = Math.max(res, helper(arr, i + 1, str));
            }

        }
        return res;
    }

    private boolean isUnique(String s) {
        Set<Character> set = new HashSet<>();
        for (char c : s.toCharArray()) {
            if (!set.add(c)) {
                return false;
            }
        }
        return true;
    }
}

File: dp/LeetCode_1239.kt
package dp

import kotlin.math.max

class LeetCode_1239 {

    private var maxLength = 0

    fun maxLength(arr: List<String>): Int {
        findMaxLength(arr, 0, "")
        return maxLength
    }

    private fun findMaxLength(arr: List<String>, index: Int, currentString: String) {
        maxLength = max(maxLength, currentString.length)

        for (i in index until arr.size) {
            if (!isUnique(currentString, arr[i])) {
                continue
            }
            findMaxLength(arr, i + 1, currentString + arr[i])
        }
    }

    private fun isUnique(currentString: String, nextString: String): Boolean {
        val uniqueArray = BooleanArray(26)

        nextString.forEach {
            if (uniqueArray[it - 'a'] || currentString.contains(it)) {
                return false
            }
            uniqueArray[it - 'a'] = true
        }

        return true
    }
}

fun main() {
    // Input
    val arr = listOf("un", "iq", "ue")
    // Output
    LeetCode_1239().maxLength(arr).run {
        println(this)
        require(4 == this)
    }
}
